### PR TITLE
Fix Get-Module -FullyQualifiedName option to work with paths

### PR DESCRIPTION
**Pull request by pougetat**
# PR Summary

This PR "fixes" the current behavior for get-module with -FullyQualifiedName specified and containing a path. Unfortunately this does mean that this PR is a breaking change.

To illustrate the current behavior and the new behavior we can look at the following example : 
/Foo/1.0/Foo.psd1
/Foo/2.0/Foo.psd1
Get-module -ListAvailable -FullyQualifiedName @{ModuleName = "/Foo"; ModuleVersion = "2.0"}

This will return both modules since "/Foo" is not correctly matched to either .psd1 file and the current behavior is to skip the version checks and simply return the modules.
With this PR the names are correctly matched and the version checks can take place thus resulting in only one module being returned. 

## PR Context

This PR is related to the following issues : 
https://github.com/PowerShell/PowerShell/issues/8936 (meta-issue tracking refactoring work on module commandlets in general)
https://github.com/PowerShell/PowerShell/issues/8262

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: https://github.com/PowerShell/PowerShell/issues/8262
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by pougetat** [comment]
CodeFactor is raising one issue (private static methods must come before non static private methods) which is not respected in other parts of the file that I have not modified. :)

File: src/System.Management.Automation/engine/Modules/GetModuleCommand.cs [hunk 1]
```diff
@@ -375,10 +375,9 @@ protected override void ProcessRecord()
             var moduleSpecTable = new Dictionary<string, ModuleSpecification>(StringComparer.OrdinalIgnoreCase);
             if (FullyQualifiedName != null)
             {
-                // TODO:
-                // FullyQualifiedName.Name could be a path, in which case it will not match module.Name.
-                // This is potentially a bug (since version checks are ignored).
-                // We should normalize FullyQualifiedName.Name here with ModuleIntrinsics.NormalizeModuleName().
+                FullyQualifiedName = FullyQualifiedName.Select(
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/Modules/GetModuleCommand.cs]: It looks like we already have some LINQ usage here, but if we can avoid it I'd like to. As nice as it is, the GC pressure it creates is a performance drain.

**Review by rjmholt — COMMENTED** [review]
@pougetat would you be able to give a concrete example of the change this makes in a comment in this PR along with an explanation of why it's a breaking change and tag the @PowerShell/powershell-committee? That way they can discuss and review it.

**Comment by pougetat** [comment]
@PowerShell/powershell-committee 
Here is an example that illustrates the behavior differences introduced with this PR.

Consider the following environment : 

/foo/1.0/foo.pds1 => version 1.0 specified in the manifest
/foo/2.0/foo.psd1 => version 2.0 specified in the manifest

When executing the following command : 

```powershell
Get-module -ListAvailable -FullyQualifiedName @{ModuleName = "./foo"; ModuleVersion = "2.0"}
```
the old behavior yields : 

```
ModuleType        Version       Name
-----------     -----------   --------
Manifest            2.0          foo
Manifest            1.0          foo
```

the new behavior yields : 
```
ModuleType        Version       Name
-----------     -----------   --------
Manifest            2.0          foo
```

This is a breaking change for anyone counting on the fact that both modules would be returned even though this behavior seems counter-intuitive.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and agree with the change in behavior as it's unlikely someone has been dependent on this existing behavior

**Review by rjmholt — APPROVED** [review]

**Comment by pougetat** [comment]
@rjmholt
Thanks for the approval and the review. I am still going to remove the extra linq expressions to not create too much garbage as you requested so this isn't quite ready to merge.

**Comment by pougetat** [comment]
@adityapatwardhan @rjmholt Hey,
I have updated the PR to remove the Linq expression :smile:

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@pougetat Thank you for your contribution!

**Comment by LaurentDardenne** [comment]
As a result of this change, as I understand it, should not the _ModuleSpecification_ class help specify that its [Name](https://github.com/PowerShell/PowerShell/blob/bd6fdae73520931f0d27a29d6290e18761772141/src/System.Management.Automation/engine/Modules/ModuleSpecification.cs#L281)  property can also contain a path and not just a module name?

**Comment by rjmholt** [comment]
> As a result of this change, as I understand it, should not the ModuleSpecification class help specify that its Name property can also contain a path and not just a module name?

Yes it should. I believe that was always intended to work, since ModuleSpecifications are how RequiredModules in PSD1s are loaded.

**Comment by rjmholt** [comment]
@LaurentDardenne which help are you talking about?

**Comment by LaurentDardenne** [comment]
> @LaurentDardenne which help are you talking about?

[This one](https://docs.microsoft.com/en-us/dotnet/api/microsoft.powershell.commands.modulespecification.name?view=pscore-6.2.0#Microsoft_PowerShell_Commands_ModuleSpecification_Name)

**Comment by rjmholt** [comment]
Oh -- that comes from comment help! (In the XML comments)

File: src/System.Management.Automation/engine/Modules/GetModuleCommand.cs [hunk 2]
```diff
@@ -545,22 +546,36 @@ private void GetLoadedModules(string[] names, IDictionary<string, ModuleSpecific
...
+        /// <param name="moduleSpecTable">The module specifications to filter candidates from.</param>
+        /// <param name="module">The module to find candidates for from the module specification table.</param>
+        /// <returns>The module specifications matching the module based on name, path and subpath.</returns>
+        private static IEnumerable<ModuleSpecification> GetCandidateModuleSpecs(
+            IDictionary<string, ModuleSpecification> moduleSpecTable,
+            PSModuleInfo module)
+        {
+            foreach (ModuleSpecification moduleSpec in moduleSpecTable.Values)
+            {
+                if (moduleSpec.Name == module.Name || moduleSpec.Name == module.Path || module.Path.Contains(moduleSpec.Name))
```
Comment (TravisEz13) [inline src/System.Management.Automation/engine/Modules/GetModuleCommand.cs]: FYI, this make existing cases of getting a module by name, using FullyQualifiedName case sensitive.  I'll submit a PR to fix this.

**Pull request merged**